fbpic/openpmd_diag/phasespace_diag.py

```diff
@@ -0,0 +1,540 @@
+
+        Ntot : int
+        	Contains the global number of particles
+
+        select_array : 1darray of bool
+            An array of the same shape as that particle array
+            containing True for the particles that satify all
+            the rules of self.select
+        """
+ 
+        # Create the datasets and set up their attributes
+        if self.rank==0:
+            datashape = self.bins
+            dtype = 'f8'
+            # If the dataset already exists, remove it.
+            # (This avoids errors with diags from previous simulations,
+            # in case the number of particles is not exactly the same.)
+            if path in species_grp:
+                del species_grp[path]
+            
+            dset = species_grp.create_dataset(path, datashape, dtype=dtype )
+
+        # Set up the sample array
+        sample = np.empty((Ntot, self.Ndims))
+    
+        # contruct the sample to be binned
+        for i in range(self.Ndims):
+            sample[:,i] = self.get_dataset( species, self.phase_space[i], 
+                                           select_array, n_rank, Ntot ) 
+            
+        # Set up a temporary list for the bin edges
+        bin_edges = []     
+        # Build the bins edges from the provided limits and deal with the
+        # moving window if required
+        if self.edges is None:
+            # If none are specified, allow histogramdd to auto-scale
+            bin_edges = None
+        else:
+            # Else, work through the provided bin edges
+            for i in range(self.Ndims):          
+                # If the phase space being binned is z, and there is a moving
+                # window set up, and the user has not specified otherwise,
+                # adjust the limits
+                if all([self.phase_space[i] == 'z', self.move_with_window, 
+                       self.comm.moving_win is not None]):
+                    t = self.comm.moving_win.t_last_move
+                    v = self.comm.moving_win.v  
+                    bin_edges.append([z+v*t for z in self.edges[i]])
+                else:
+                    # Otherwise use the edges as they are
+                    bin_edges.append(self.edges[i])
+
+        # Get the particle weights
+        weights = self.get_dataset( species, 'w', select_array,
+                                           n_rank, Ntot )
+        
+        # deposit a specified quantity alongside the weights
+        if self.deposit != 'w':
+            deposit = self.get_dataset( species, self.deposit, select_array, 
+                                        n_rank, Ntot )            
+            weights *= deposit
+            # set the deposition metadata
+            dset.attrs['deposit'] = self.deposit 
+        else:
+            dset.attrs['deposit'] = 'weight'        
+  
+        # generate the histogram
+        hist, edges = np.histogramdd(sample, bins=self.bins, 
```

Comment:
I would suggest to also put this call to `np.histogramdd` in the `if self.rank == 0:` condition below:
only the MPI rank that writes the results needs to perform this calculation.